### PR TITLE
Allow device transfer when graph is connected

### PR DESCRIPTION
**Pull request by hvy**
Fixes https://github.com/chainer/chainer/issues/7310 (in particular https://github.com/chainer/chainer/issues/7310#issuecomment-501577567).

- [x] Allow device transfer.
- [x] Test allowed device transfer.
- [x] Unchain.
- [x] Test unchain.

### CODE REVIEW

**Comment by toslunar** [comment]
I'm for the design in that `to_device` unchains the graph even if the destination is the current device.

Unchaining a non-trivial graph could be warned.  Otherwise it would become harder to find mistakes like calling `some_stateful_link.to_device` every iteration.  `no_backprop_mode` may suppress the warning.

**Comment by hvy** [comment]
I'm starting to feel that the `Variable` class is difficult to work with. There are so many states (attributes) that are managed by customized properties, calling private methods, etc.

**Comment by hvy** [comment]
@toslunar thanks for the comment. Could you explain a bit more in what cases that would be an issue? Already in the current master, if you transfer the device of some variable in the chain, you'll most likely get an error during backprop that devices mismatch (of inputs to `FunctionNode`s).

File: chainer/variable.py [hunk 1]
```diff
@@ -1134,96 +1131,98 @@ def from_chx(self):
...
+        if arr is not None:
+            if was_chainerx:
+                arr = arr.as_grad_stopped(copy=False)
+            new_arr = device.send(arr)
+            if is_chainerx and self.requires_grad:
+                new_arr.require_grad()
+
+                # Gradient is updated torwards the end of this method and is
+                # therefore passed as `None`.
+                self._set_chainerx_array(new_arr, None)
```
Comment (hvy) [inline chainer/variable.py]: `_set_chainerx_array` don't really need the `grad` argument?

**Comment by toslunar** [comment]
> mistakes like calling `some_stateful_link.to_device` every iteration.

For example, the PTB example with the following change will cut the graph like `--bproplen=1`, but it runs.  It seems `Optimizer.setup` before `Link.to_device` is allowed.
```diff
diff --git a/examples/ptb/train_ptb_custom_loop.py b/examples/ptb/train_ptb_custom_loop.py
index 86b1a23..00eb7b1 100755
--- a/examples/ptb/train_ptb_custom_loop.py
+++ b/examples/ptb/train_ptb_custom_loop.py
@@ -101,7 +101,6 @@ def main():
     rnn = train_ptb.RNNForLM(n_vocab, args.unit)
     model = L.Classifier(rnn)
     model.compute_accuracy = False  # we only want the perplexity
-    model.to_device(device)

     # Set up an optimizer
     optimizer = chainer.optimizers.SGD(lr=1.0)
@@ -136,7 +135,8 @@ def main():
             # (it is chainer.dataset.concat_examples by default)
             x, t = convert.concat_examples(batch, device)
             # Compute the loss at this time step and accumulate it
-            loss += optimizer.target(x, t)
+            loss += model.to_device(device)(x, t)
+
             count += 1

         sum_perp += loss.array
```
```
~/chainer/examples/ptb$ ./train_ptb_custom_loop.py --test --device=@cupy:0
#vocab = 10000
epoch: 7
validation perplexity: 6360.183448360342
epoch: 14
validation perplexity: 756241.2535119326
epoch: 21
validation perplexity: 4625.535995634493
epoch: 28
validation perplexity: 289800.56027632434
epoch: 35
validation perplexity: 11076.9411158485
epoch: 42
validation perplexity: 69745.10646834601
test
test perplexity: 30384.287894353685
save the model
save the optimizer
```

**Comment by hvy** [comment]
It seems like a rather rare case but please correct me. Is it not okay document the behavior? 

However, I think it's good to show a warning if it's possible to do it in a clean way. Did you have something in mind on how that could be implemented?

**Comment by hvy** [comment]
How about the following?
- Always unchain.
- Raise a warning when transferring a variable that is not a leaf to the same device. This would be to alert the user writing the above code provided by @toslunar .

**Comment by hvy** [comment]
The current status of the discussion is 
- Always unchain.
- Introduce an argument to control whether unchaining is allowed.
- Raise a `DeprecationWarning` if a transfer would result in an unchaining operation (variable it not leaf) while the argument is not specified to allow unchaining. In the future, remove the warning and raise an error.

The motivation is that it's better to let the user be explicit about unchaining to avoid unexpected behavior. However, we do not want to start raising errors immediately since user code will start to break, especially, we'd like to avoid user code to start failing post training but pre serialization, for instance.

**Comment by hvy** [comment]
Added an argument to control whether unchaining may occur or not. PTAL. I'll add tests if this looks okay.

**Comment by hvy** [comment]
Need to fix the DeviceResident's device transfer methods as well so that Links will behaive similar to the Variable. Also, it should be better to use a less misleading word for the argument such as `allow_unchain` (`unchain=False` will raise an error but it sounds like it should transfer the device, but without unchaining).

**Comment by hvy** [comment]
I want to add test for the Variable tests also but would like to refactor the relevant tests before that.

**Comment by hvy** [comment]
I'll try to clean up and add more Variable tests.

**Comment by hvy** [comment]
Refactor to_device-family tests in test_variable. PTAL.

File: chainer/device_resident.py [hunk 2]
```diff
@@ -56,28 +56,39 @@ def xp(self):
...
+    ):
+        # type: (...) -> 'DeviceResident'
         """Copies parameter variables and persistent values to CPU.
 
         This method does not handle non-registered attributes. If some of such
         attributes must be copied to CPU, the link implementation should
         override :meth:`~DeviceResident.device_resident_accept` to do so.
 
+        Args:
+            allow_unchain (bool): If ``True``, always unchains the graph. If
```
Comment (niboshi) [inline chainer/device_resident.py]: Is the last spec (`None`) needed? The default can simply be `False`.

File: chainer/device_resident.py [hunk 3]
```diff
@@ -56,28 +56,39 @@ def xp(self):
...
+                ``False``, stops and raises an error in case a graph would be
```
Comment (niboshi) [inline chainer/device_resident.py]: * "stops and" can be removed (readers would wonder whether it has some special meaning).
* "in case a graph would be unchained.": How about the opposite logic e.g.  "in case the link was connected to a graph" or something? "would be unchained" has some ambiguity: readers might think it applies to the case where the variables were already unchained (especially those who are not very good at English).

File: chainer/device_resident.py (hunk 2 above)
Comment (hvy) [reply chainer/device_resident.py]: If the default value is `False`, existing user code will break. That's what we ultimately want to but only after the deprecation phase.

Comment (niboshi) [reply chainer/device_resident.py]: How about simply raising a warning instead of an error in the deprecation period?

Comment (hvy) [reply chainer/device_resident.py]: I see that's another optional. Could you explain why that's better? It sounds like the user has to be more aware of when the deprecation ends.

Comment (niboshi) [reply chainer/device_resident.py]: Of course affected users should be aware of the change. You don't think so?

Another point is simplicity. It's difficult for users to understand the necessity of the third value.

Comment (hvy) [reply chainer/device_resident.py]: I agree on the simplicity. Let me change it. The deprecation warning is only raised when the graphs are connected so it should be rather rare too and not too disturbing neither.

Comment (hvy) [reply chainer/device_resident.py]: Changed the behavior. Also need to fix the docs.

Comment (hvy) [reply chainer/device_resident.py]: Fixed docs.

File: chainer/device_resident.py (hunk 3 above)
Comment (hvy) [reply chainer/device_resident.py]: Fixed the docs.

**Comment by hvy** [comment]
PTAL.

File: chainer/variable.py [hunk 4]
```diff
@@ -1088,158 +1088,232 @@ def T(self):
...
+        # - ChainerX -> ChainerX: Calling `as_grad_stopped`.
 
-        if arr is None:
-            return
+        if not was_chainerx:
+            self.unchain()
 
-        if backend.get_device_from_array(arr) == device:
-            return
+        # Transfer gradient variable if any.
```
Comment (niboshi) [inline chainer/variable.py]: Hmm, could you explain why the code is restructured this way (send grad_var -> send data -> set grad_var)?

File: chainer/device_resident.py [hunk 5]
```diff
@@ -56,28 +56,44 @@ def xp(self):
...
-            starting_device_resident=self)
+            starting_device_resident=self,
+            allow_unchain=allow_unchain)
         self.__to_device(visitor)
         return self
 
     def to_gpu(
             self,
             device=None,  # type: tp.Optional[types.CudaDeviceSpec]
+            allow_unchain=False,  # type: tp.Optional[bool]
```
[the remark this answers is not in the document]
Comment (hvy) [reply chainer/device_resident.py]: Fixed.

File: chainer/device_resident.py [hunk 6]
```diff
@@ -86,9 +102,18 @@ def to_gpu(
         attributes must be copied to GPU, the link implementation must
         override :meth:`~DeviceResident.device_resident_accept` to do so.
 
+        .. note::
+
+            This method will raise an error in the future instead of a warning
+            when the object is connected to a graph but `allow_unchain` is not
```
[the remark this answers is not in the document]
Comment (hvy) [reply chainer/device_resident.py]: Thanks, fixed.

File: chainer/variable.py (hunk 4 above)
Comment (hvy) [reply chainer/variable.py]: It's because `_set_chainerx_array` also modifies the gradient and passing the gradient to this method is rather tricky during device transfer (that's why it's done outside this method). IMO, this method can actually be simplified not to take the gradient argument at all.

Comment (niboshi) [reply chainer/variable.py]: I'd like you to explain this part a bit more:
>passing the gradient to this method is rather tricky during device transfer

Comment (hvy) [reply chainer/variable.py]: Sorry let me rephrase it. I'm not sure how to do that in a clean way that's simpler than the current implementation. Do you have any ideas?

Comment (niboshi) [reply chainer/variable.py]: I'm not saying which is better.
I'm simply asking you to explain the reason of the change.

Comment (hvy) [reply chainer/variable.py]: We now have to unchain `chainerx.ndarrays`, which we didn't have to before. This was basically the only way I could make it work. Internal methods such as `_set_chainerx_array`, `{_to,to}_device`, `grad_var` custom setter, etc. all highly depend on the internal state of the variable. Modifying the state of the variable without breaking it is non-trivial.

Comment (niboshi) [reply chainer/variable.py]: Thanks, I have been asking that kind of information.

I don't still understand what you meant by `_set_chainerx_array` depends on the internal state. In my understanding the method is meant to be just a dumb assigner. The internal state should be handled by the caller.

Comment (hvy) [reply chainer/variable.py]: It may be dumb but it depends on `self. _requires_grad` and modifies `self._has_chainerx_array` which has a huge effect on the behavior of the variable as most attribute custom getters checks this variable.

Comment (niboshi) [reply chainer/variable.py]: OK. I think the method should accept `requires_grad` as an agrument (in fact it was originally implemented that way). Do you think that could make the code simpler?

Comment (hvy) [reply chainer/variable.py]: That a might be a better design choice of `_has_chainerx_array` (it'll also be more clean when `grad` isn't used) but that alone won't help much for this PR (I'd like to use `Variable.to_device` for the gradient too, and not this internal method, since it's simpler). I think it's better to reconsider it at a Variable-wide scope if it should be refactored.

Comment (niboshi) [reply chainer/variable.py]: I'm beginning to understand.

>I'd like to use `Variable.to_device` for the gradient too

This was actually the most important point of the refactoring and the change of the flow, right?
It would have been nice if you had revealed the intent at the very beginning.

What if `_set_chainerx_array` have `grad_var` as an argument?

Comment (niboshi) [reply chainer/variable.py]: Perhaps I can merge it in the current status and consider refactoring later.

Comment (hvy) [reply chainer/variable.py]: I'm definitely for refactoring the Variable internals for the sake of maintainability and also to do it separately outside this PR. It should have more time and though put into it.

Comment (niboshi) [reply chainer/variable.py]: OK, I agree.

File: tests/chainer_tests/links_tests/connection_tests/test_lstm.py [hunk 7]
```diff
@@ -184,11 +184,14 @@ def test_to_cpu_gpu(self):
         self.check_to_cpu(self.link.h)
 
     def check_to_cpu_to_gpu(self, s):
-        self.link.to_gpu()
+        # If `self.link` already resides on the GPU, the first `to_gpu` calls
+        # do nothing and the first device transfer occurs at the first
+        # `to_cpu`. We must therefore pass `allow_unchain=True` there.
+        self.link.to_gpu(allow_unchain=True)
```
[the remark this answers is not in the document]
Comment (hvy) [reply tests/chainer_tests/links_tests/connection_tests/test_lstm.py]: This is not necessarily a GPU-GPU transfer https://github.com/chainer/chainer/pull/7507/files#diff-c27c7baca59d10196efd36f373e887fbR202

[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/links_tests/connection_tests/test_lstm.py]: Could you give more explanation?
I don't understand the logical association between the two sentences.

[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/links_tests/connection_tests/test_lstm.py]: Btw, the link does not seem to work.

[the remark this answers is not in the document]
Comment (hvy) [reply tests/chainer_tests/links_tests/connection_tests/test_lstm.py]: Updated the link.

[the remark this answers is not in the document]
Comment (hvy) [reply tests/chainer_tests/links_tests/connection_tests/test_lstm.py]: Do you mean the two sentences in the comment in the code above this line? If yes, when this line is executed
- if the link already resides on the GPU, it essentially does nothing, not even unchain. So, we must again when transferring to CPU further down this check method, again pass `allow_unchain=True` to avoid the deprecation error
- if on the other hand the link resides on the CPU, then obviously this list must look like this in order to avoid the deprecation error.

[the remark this answers is not in the document]
Comment (niboshi) [reply tests/chainer_tests/links_tests/connection_tests/test_lstm.py]: Let's fix so that `allow_unchain=True` takes effect in GPU-GPU case.

**Comment by hvy** [comment]
Fixed skip-between-cupy-device transfers to also unchain.

**Comment by hvy** [comment]
PTAL.

**Comment by niboshi** [comment]
Please resolve conflict.

**Comment by hvy** [comment]
Fixed.

File: tests/chainer_tests/test_variable.py [hunk 8]
```diff
@@ -983,358 +977,314 @@ def v():
...
+        self.gx = gx
 
-    @attr.chainerx
-    def test_to_cpu_from_chainerx_requiring_grad(self):
-        with self.assertRaises(RuntimeError):
-            self.check_to_cpu(
-                chainerx.array(self.x),
-                chainerx.array(self.gx),
-                requires_grad=True)
+    def to_device_meth(self, dst_device, allow_unchain):
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: Let's spell out till the end.
(ditto for all others)

```suggestion
    def to_device_method(self, dst_device, allow_unchain):
```

File: tests/chainer_tests/test_variable.py [hunk 9]
```diff
@@ -983,358 +977,314 @@ def v():
...
+        # Call `self.x.to_device`, `self.x.from_chx`, etc.
+        raise NotImplementedError('Must be overridden.')
 
+    def check_to_device_meth_devices(self, src_device, dst_device):
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: The suffix `_devices` does not sound representing this method.

File: tests/chainer_tests/test_variable.py [hunk 10]
```diff
@@ -983,358 +977,314 @@ def v():
...
+        # Called after the device transfer, check `self.x`, etc.
+        raise NotImplementedError('Must be overridden.')
 
-@testing.parameterize(
-    {'x_shape': (10,)},
-    {'x_shape': ()},
-)
-@attr.gpu
-class TestVariableToGpu(unittest.TestCase):
+    def check_to_device_meth_ndarrays(self):
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: Please write a comment to this method too.
The distinction from `check_to_device_meth_devices` is not clear.

File: tests/chainer_tests/test_variable.py [hunk 11]
```diff
@@ -983,358 +977,314 @@ def v():
             assert v().requires_grad is self.expected
 
 
-@testing.parameterize(
-    {'x_shape': (10,)},
-    {'x_shape': ()},
-)
-class TestVariableToCpu(unittest.TestCase):
+# TODO(hvy): Classes deriving this base class should test all backends
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: Is this TODO?
Each subclass has two backend decorators. I don't understand which this comment is referring to.

File: tests/chainer_tests/test_variable.py [hunk 12]
```diff
@@ -983,358 +977,314 @@ def v():
...
+@testing.backend.inject_backend_tests(None, _backend_params)
+class TestVariableToCpu(VariableToDeviceFamilyBase, unittest.TestCase):
 
-@testing.parameterize(
-    {'x_shape': (10,)},
-    {'x_shape': ()},
-)
-@attr.chainerx
-class TestVariableToChainerX(unittest.TestCase):
+    def to_device_meth(self, _, allow_unchain):
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: Let's not replace the unused argument with `_`.
It's not clear what it was and that it's really something to ignore.
Ditto for others.

File: tests/chainer_tests/test_variable.py (hunk 11 above)
Comment (hvy) [reply tests/chainer_tests/test_variable.py]: ideep is not tested at all. I'll remove the comment if that's okay.

Comment (niboshi) [reply tests/chainer_tests/test_variable.py]: In that case it should be a TODO.
Please be more specific about what to be done.
I think it's better to put a TODO just before each line that needs to be fixed.

File: tests/chainer_tests/test_variable.py (hunk 8 above)
Comment (hvy) [reply tests/chainer_tests/test_variable.py]: Fixed.

File: tests/chainer_tests/test_variable.py (hunk 10 above)
Comment (hvy) [reply tests/chainer_tests/test_variable.py]: Commented. What do you think?

File: tests/chainer_tests/test_variable.py (hunk 12 above)
Comment (hvy) [reply tests/chainer_tests/test_variable.py]: Fixed!

File: tests/chainer_tests/test_variable.py (hunk 11 above)
Comment (hvy) [reply tests/chainer_tests/test_variable.py]: That sounds good, changed accordingly.

File: tests/chainer_tests/test_variable.py (hunk 10 above)
Comment (niboshi) [reply tests/chainer_tests/test_variable.py]: Looks good!

File: tests/chainer_tests/test_variable.py [hunk 13]
```diff
@@ -983,358 +977,320 @@ def v():
...
-    def test_ideep_to_chx(self):
-        raise unittest.SkipTest('Not yet supported')
+@testing.parameterize(*testing.product({
+    'x_shape': [(10,), ()],
+    'x_with_grad': [True, False],
+    'x_with_graph': [True, False],
+}))
+@testing.backend.inject_backend_tests(None, _numpy_backend_params)
+# TODO(hvy): Test intel64.
+@testing.backend.inject_backend_tests(None, _backend_params)
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: Hmm, how about adding a TODO at the definition of `_backend_params` instead?
These lines themselves do not seem to need fixes.

Comment (niboshi) [reply tests/chainer_tests/test_variable.py]: Is there any reason intel64 cannot be tested for now?

Comment (hvy) [reply tests/chainer_tests/test_variable.py]: The reason is `to_chx`, which won't work on intel64 ndarray holding variables (`RuntimeError`). It is probably possible for the other methods though.

Comment (niboshi) [reply tests/chainer_tests/test_variable.py]: OK, please use skip.

Comment (hvy) [reply tests/chainer_tests/test_variable.py]: Needed some changes/skips to the tests but fixed to also test intel64.

File: tests/chainer_tests/functions_tests/pooling_tests/test_upsampling_2d.py [hunk 14]
```diff
@@ -55,12 +55,12 @@ def check_forward(self, y):
                 testing.assert_allclose(up_y, 0)
 
     def test_forward_cpu(self):
-        self.pooled_y.to_cpu()
+        self.pooled_y.to_cpu(allow_unchain=True)
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/pooling_tests/test_upsampling_2d.py]: I think this `allow_unchain=True` is undesirable, because it changes the state of `pooled_y` and it's nonessential for the test. How about calling `self.pooled_y.unchain()` in `setUp` (or TODO)?

File: tests/chainer_tests/functions_tests/pooling_tests/test_upsampling_2d.py [hunk 15]
```diff
@@ -55,12 +55,12 @@ def check_forward(self, y):
...
         self.check_forward(self.pooled_y)
 
     @attr.gpu
     def test_forward_gpu(self):
-        self.pooled_y.to_gpu()
+        self.pooled_y.to_gpu(allow_unchain=True)
```
Comment (niboshi) [inline tests/chainer_tests/functions_tests/pooling_tests/test_upsampling_2d.py]: ditto.

File: tests/chainer_tests/test_variable.py [hunk 16]
```diff
@@ -983,358 +981,325 @@ def v():
...
 
-    @attr.chainerx
-    def test_to_cpu_from_chainerx_requiring_grad(self):
-        with self.assertRaises(RuntimeError):
-            self.check_to_cpu(
-                chainerx.array(self.x),
-                chainerx.array(self.gx),
-                requires_grad=True)
+        if self.x_with_graph:
+            x *= x
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: This is a bit difficult to understand. How about treating intel64 backend a special case, or at least leave some comment.

File: tests/chainer_tests/functions_tests/pooling_tests/test_upsampling_2d.py (hunk 14 above)
Comment (hvy) [reply tests/chainer_tests/functions_tests/pooling_tests/test_upsampling_2d.py]: Okay, fixed to call `unchain` in `setUp`.

File: tests/chainer_tests/functions_tests/pooling_tests/test_upsampling_2d.py (hunk 15 above)
Comment (hvy) [reply tests/chainer_tests/functions_tests/pooling_tests/test_upsampling_2d.py]: Fixed.

File: tests/chainer_tests/test_variable.py (hunk 16 above)
Comment (hvy) [reply tests/chainer_tests/test_variable.py]: Added a comment.

**Comment by hvy** [comment]
PTAL.

File: tests/chainer_tests/test_variable.py (hunk 16 above)
Comment (niboshi) [reply tests/chainer_tests/test_variable.py]: Well, the original code was `x = 2 * x`. How about writing that? In-place operation is confusing here, in the first place.
By the way, Is "python scalar" essential? If `x = x * x` works, how about using that?

Comment (hvy) [reply tests/chainer_tests/test_variable.py]: Changed to a reassignment.

Comment (hvy) [reply tests/chainer_tests/test_variable.py]: I won't say that an example is essential. However, it makes the comment a lot clearer demonstrating to what concrete types `x` may be cast and under what circumstances.

Comment (niboshi) [reply tests/chainer_tests/test_variable.py]: Hm, if `x = x * x` works, I don't even think the comment is necessary at all.
What do you think about that?

Comment (hvy) [reply tests/chainer_tests/test_variable.py]: I'm fine either way. I can remove it. I must say that this behavior was a surprise to me though.

Comment (hvy) [reply tests/chainer_tests/test_variable.py]: Removed.

Comment (niboshi) [reply tests/chainer_tests/test_variable.py]: >I must say that this behavior was a surprise to me though.

Yes, but it's no longer relevant to this test code.

File: tests/chainer_tests/test_variable.py [hunk 17]
```diff
@@ -983,358 +981,325 @@ def v():
...
 
-    @attr.multi_gpu(2)
-    def test_to_gpu_from_another_gpu(self):
-        self.check_to_gpu(cuda.to_gpu(self.x), cuda.to_gpu(self.gx), 1)
+        if self.x_with_graph:
+            with pytest.warns(DeprecationWarning):
+                self.to_device_method(dst_backend_config.device, False)
+        else:
+            # Check that nothing is warned.
+            self.to_device_method(dst_backend_config.device, False)
```
Comment (niboshi) [inline tests/chainer_tests/test_variable.py]: It seems it does not check as written in the comment?

Comment (hvy) [reply tests/chainer_tests/test_variable.py]: Hm, it checks (do nothing if it's okay, else raise) just not explicitly but by not wrapping it in an expect-warning context. I think this comment makes sense when you look at his this method as a whole (in particular https://github.com/chainer/chainer/pull/7507/files/757e1b344e5222d43a86377386a247a838887db0#diff-278878af6cd00cabfd7437aa9efff4f2R1062).

Comment (niboshi) [reply tests/chainer_tests/test_variable.py]: Is that true? I think warnings do not make tests failure?

Comment (hvy) [reply tests/chainer_tests/test_variable.py]: Yes, they are configured in `setup.cfg` and deprecation warnings in particular are treated as errors https://github.com/chainer/chainer/blob/master/setup.cfg#L12.

Comment (niboshi) [reply tests/chainer_tests/test_variable.py]: Ah, I see.
I think "nothing is warned" is misleading: only specific warning types are checked.
By the way, `setup.cfg` is not the test itself: it only defines how tests are run. It's better not rely on it this way.

You can use `pytest.warns`: https://stackoverflow.com/a/45671804

Comment (hvy) [reply tests/chainer_tests/test_variable.py]: Fixed to check the warning count to check cases when no unchaining takes place.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Please check test results.

**Comment by hvy** [comment]
They should be fixed now, PTAL.

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
This error in example test seems persistent:

```
11:44:45 Traceback (most recent call last):
11:44:45   File "examples/mnist/train_mnist_model_parallel.py", line 107, in <module>
11:44:45     main()
11:44:45   File "examples/mnist/train_mnist_model_parallel.py", line 103, in main
11:44:45     trainer.run()
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/training/trainer.py", line 349, in run
11:44:45     six.reraise(*exc_info)
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/training/trainer.py", line 316, in run
11:44:45     update()
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/training/updaters/standard_updater.py", line 205, in update
11:44:45     self.update_core()
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/training/updaters/standard_updater.py", line 218, in update_core
11:44:45     optimizer.update(loss_func, *in_arrays)
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/optimizer.py", line 859, in update
11:44:45     loss = lossfun(*args, **kwds)
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/link.py", line 294, in __call__
11:44:45     out = forward(*args, **kwargs)
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/links/model/classifier.py", line 143, in forward
11:44:45     self.y = self.predictor(*args, **kwargs)
11:44:45   File "examples/mnist/train_mnist_model_parallel.py", line 36, in __call__
11:44:45     z1 = self.first1(x1)
11:44:45   File "/work/chainer-2.0.0/examples/mnist/train_mnist.py", line 32, in __call__
11:44:45     h1 = F.relu(self.l1(x))
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/link.py", line 294, in __call__
11:44:45     out = forward(*args, **kwargs)
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/links/connection/linear.py", line 155, in forward
11:44:45     return linear.linear(x, self.W, self.b, n_batch_axes=n_batch_axes)
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/functions/connection/linear.py", line 308, in linear
11:44:45     y, = LinearFunction().apply(args)
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/function_node.py", line 325, in apply
11:44:45     outputs = self.forward(in_data)
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/functions/connection/linear.py", line 118, in forward
11:44:45     self.static_add_bias(inputs=[b], outputs=[y])
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/graph_optimizations/static_graph_utilities.py", line 186, in wrapped_func
11:44:45     ret = func(*args, **kwargs)
11:44:45   File "/home/user/.local/lib/python2.7/site-packages/chainer/functions/connection/linear.py", line 64, in static_add_bias
11:44:45     y += bias
11:44:45   File "cupy/core/core.pyx", line 1062, in cupy.core.core.ndarray.__iadd__
11:44:45   File "cupy/core/_kernel.pyx", line 851, in cupy.core._kernel.ufunc.__call__
11:44:45   File "cupy/core/_kernel.pyx", line 874, in cupy.core._kernel.ufunc._get_ufunc_kernel
11:44:45   File "cupy/core/_kernel.pyx", line 625, in cupy.core._kernel._get_ufunc_kernel
11:44:45   File "cupy/core/_kernel.pyx", line 60, in cupy.core._kernel._get_simple_elementwise_kernel
11:44:45   File "cupy/core/carray.pxi", line 166, in cupy.core.core.compile_with_cache
11:44:45   File "/work/cupy/cupy/cuda/compiler.py", line 162, in compile_with_cache
11:44:45     mod.load(cubin)
11:44:45   File "cupy/cuda/function.pyx", line 184, in cupy.cuda.function.Module.load
11:44:45   File "cupy/cuda/function.pyx", line 186, in cupy.cuda.function.Module.load
11:44:45   File "cupy/cuda/driver.pyx", line 186, in cupy.cuda.driver.moduleLoadData
11:44:45   File "cupy/cuda/driver.pyx", line 82, in cupy.cuda.driver.check_status
11:44:45 cupy.cuda.driver.CUDADriverError: CUDA_ERROR_ILLEGAL_ADDRESS: an illegal memory access was encountered
```

**Comment by niboshi** [comment]
Jenkins, test this please

**Comment by niboshi** [comment]
>This error in example test seems persistent:

Seems so.

**Comment by niboshi** [comment]
Just notifying this PR requires a fix.
